File: OpenMPI/mpi-gametheory.c
#include <stdlib.h>
#include <stdio.h>
#include <mpi.h>

typedef char Bool;
const char False = 0; 
const char True = 1; 

typedef struct Player {
	float   probOfCooperateIfOppoCompeted;
	float   probOfCooperateIfOppoCooperated;
	Bool    iCooperatedLastMove;
	Bool    oppoCooperatedLastMove;
} Player;

const unsigned int PLAYER_A_INDEX = 0;
const unsigned int PLAYER_B_INDEX = 1;

void calculatePayoffs( Bool playerCooperations[2], float payoffResults[2] ) {
	if ( playerCooperations[PLAYER_A_INDEX] && playerCooperations[PLAYER_B_INDEX] ) {
		payoffResults[PLAYER_A_INDEX] = 2;
		payoffResults[PLAYER_B_INDEX] = 2;
	}
	else if ( !playerCooperations[PLAYER_A_INDEX] && !playerCooperations[PLAYER_B_INDEX] ) {
		payoffResults[PLAYER_A_INDEX] = 1;
		payoffResults[PLAYER_B_INDEX] = 1;
	}
	else if ( !playerCooperations[PLAYER_A_INDEX] && playerCooperations[PLAYER_B_INDEX] ) {
		payoffResults[PLAYER_A_INDEX] = 3;
		payoffResults[PLAYER_B_INDEX] = 0;
	}
	else if ( playerCooperations[PLAYER_A_INDEX] && !playerCooperations[PLAYER_B_INDEX] ) {
		payoffResults[PLAYER_A_INDEX] = 0;
		payoffResults[PLAYER_B_INDEX] = 3;
	}
}

Bool player_GetDecision( Player* player, int randomPercent ) {
	float percentageProbCooperate;

	if ( player->oppoCooperatedLastMove ) {
		percentageProbCooperate = 100 * player->probOfCooperateIfOppoCooperated;
	}
	else {
		percentageProbCooperate = 100 * player->probOfCooperateIfOppoCompeted;
	}
	if ( randomPercent < percentageProbCooperate ) {
		return True;
	}
	else {
		return False;
	}
}


int main( int argc, char** argv ) {
	Player thisPlayer;
	float playerSums[2] = { 0, 0 };
	float payoffResults[2];
	Bool playerCooperations[2];
	int randomPercentages[2][100];
	int totalRounds = 100;
	unsigned int round_I = 0;
	int myRank = 0;
	unsigned int thisPlayerIndex = 0;
	unsigned int otherPlayerIndex = 0;
	MPI_Request request;
	MPI_Status recvStatus;

	MPI_Init( &argc, &argv );
	MPI_Comm_rank( MPI_COMM_WORLD, &myRank );
	
	/* create the players */
	if ( myRank == PLAYER_A_INDEX ) {
		/* Make player A "nice" */
		thisPlayer.probOfCooperateIfOppoCompeted = 0.5;
		thisPlayer.probOfCooperateIfOppoCooperated = 1;
		thisPlayer.iCooperatedLastMove = True;
		thisPlayer.oppoCooperatedLastMove = True;

		thisPlayerIndex = PLAYER_A_INDEX;
		otherPlayerIndex = PLAYER_B_INDEX;
	}
	else {
		/* Make player B "sneaky" */
		thisPlayer.probOfCooperateIfOppoCompeted = 0.25;
		thisPlayer.probOfCooperateIfOppoCooperated = 0.7;
		thisPlayer.iCooperatedLastMove = True;
		thisPlayer.oppoCooperatedLastMove = True;

		thisPlayerIndex = PLAYER_B_INDEX;
		otherPlayerIndex = PLAYER_A_INDEX;
	}

	/* Pre-Generate the percentages for each round, in order to be repeatable */
	srand( 1 );
	for ( round_I = 0; round_I < totalRounds; round_I++ ) {
		randomPercentages[PLAYER_A_INDEX][round_I] = rand() % 100;
		randomPercentages[PLAYER_B_INDEX][round_I] = rand() % 100;
	}
	
	for ( round_I = 0; round_I < totalRounds; round_I++ ) {
		/* Calculate current player's decision */
		playerCooperations[thisPlayerIndex] = player_GetDecision(
			&thisPlayer, randomPercentages[thisPlayerIndex][round_I] );

		/* recv other player's move into playerCooperations[otherPlayerIndex] */
		MPI_Irecv( &playerCooperations[otherPlayerIndex], 1, MPI_CHAR, otherPlayerIndex,
			0, MPI_COMM_WORLD, &request );

		/* send my move to other player */
		MPI_Send( &playerCooperations[thisPlayerIndex], 1, MPI_CHAR, otherPlayerIndex,
			0, MPI_COMM_WORLD );
		MPI_Wait( &request, &recvStatus );

		/* Inform player A of player B's decision */
		thisPlayer.oppoCooperatedLastMove = playerCooperations[otherPlayerIndex];

		/* Calculate the payoffs */
		calculatePayoffs( playerCooperations, payoffResults );
		
		/* update the sums */
		playerSums[thisPlayerIndex] += payoffResults[thisPlayerIndex];
		playerSums[otherPlayerIndex] += payoffResults[otherPlayerIndex];
	}

	printf( "Proc %u: After %d rounds, player A has earnt $%f, player B has earnt $%f\n",
		myRank, totalRounds, playerSums[PLAYER_A_INDEX], playerSums[PLAYER_B_INDEX] );
		
	MPI_Finalize();
	return 0;	
}

File: OpenMPI/mpi-gametheory.f90
        program mainFunction 

          implicit none

          include 'mpif.h'
          ! player A is first entry in arrays
          ! player B is second entry
	  real, dimension(2)   :: probCoopIfOppComp
          real, dimension(2)   :: probCoopIfOppCoop
	  logical,dimension(2) :: oppCoopLastMove
 
          integer, parameter :: playerAIndex=1
          integer, parameter :: playerBIndex=2
          integer, parameter :: totalRounds=100

	  real          :: playerA_Sum, playerB_Sum
          integer, dimension(2,100) :: randomPercentages(2, 100)
          real, dimension(2)        :: payoffResults
          logical, dimension(2)     :: playerCoops

	  logical       :: playerDec
          integer       :: round_I
          integer       :: thisPlayerIndex, otherPlayerIndex
          
          integer       :: rq
          integer       :: st(MPI_STATUS_SIZE)
          integer       :: rank, ierr

          call MPI_INIT( ierr )
          call MPI_COMM_RANK( MPI_COMM_WORLD, rank, ierr )

          round_I = 0
          playerA_Sum = 0
          playerB_Sum = 0
          
          ! create the players 
          if( rank .eq. 0 ) then 
          ! Make player A "nice"
            probCoopIfOppComp(playerAIndex) = 0.5
            probCoopIfOppCoop(playerAIndex) = 1
            playerCoops(playerAIndex) = .True.
            oppCoopLastMove(playerAIndex) = .True.

            thisPlayerIndex = playerAIndex
            otherPlayerIndex = playerBIndex
          else 
          ! Make player B "sneaky" 
            probCoopIfOppComp(playerBIndex) = 0.25
            probCoopIfOppCoop(playerBIndex) = 0.7
            playerCoops(playerBIndex) = .True.
            oppCoopLastMove(playerBIndex) = .True.

            thisPlayerIndex = playerBIndex
            otherPlayerIndex = playerAIndex
          end if

          ! Pre-Generate the percentages for each round, in order to be repeatable 
          call SRAND( 1 )
          do round_I = 0, totalRounds-1
            randomPercentages(playerAIndex, round_I) = rand() * 100
            randomPercentages(playerBIndex, round_I) = rand() * 100
          end do
         
          do round_I = 0, totalRounds-1
            ! query this player decision
            playerCoops(thisPlayerIndex) = &
            &    playerDec(oppCoopLastMove(thisPlayerIndex), &
            &    probCoopIfOppCoop(thisPlayerIndex), &
            &    probCoopIfOppComp(thisPlayerIndex), &
            &    randomPercentages(thisPlayerIndex, round_I) )

            call MPI_IRECV( playerCoops(otherPlayerIndex), 1, &
            &               MPI_LOGICAL, otherPlayerIndex-1, 0, &
            &               MPI_COMM_WORLD, rq, ierr  )
            call MPI_SEND( playerCoops(thisPlayerIndex), 1, &
            &              MPI_LOGICAL, otherPlayerIndex-1, 0, &
            &              MPI_COMM_WORLD, ierr )
            call MPI_WAIT( rq, st, ierr )
            
            ! inform this player of other player's decision
            oppCoopLastMove(thisPlayerIndex) = playerCoops(otherPlayerIndex)

            ! Calculate the payoffs 
            call calculatePayoffs( playerCoops, payoffResults )

            ! update the sums 
            playerA_Sum = playerA_Sum + payoffResults(playerAIndex)
            playerB_Sum = playerB_Sum + payoffResults(playerBIndex)
          end do

          print*, rank, "After ",totalRounds, &
          &       " rounds, player A has earnt $", &
          &       playerA_Sum,", player B has earnt $", playerB_Sum 

          call MPI_FINALIZE( ierr )

        end program

        subroutine calculatePayoffs( playerCoops, payoffResults )
          logical, dimension(2) :: playerCoops
          real, dimension(2)    :: payoffResults

          integer, parameter    :: playerAIndex=1
          integer, parameter    :: playerBIndex=2

          if( playerCoops(playerAIndex) .and. &
            & playerCoops(playerBIndex) ) then
            payoffResults(playerAIndex) = 2
            payoffResults(playerBIndex) = 2
          else if( (.not.playerCoops(playerAIndex)) .and. &
          &        (.not.playerCoops(playerBIndex))) then
            payoffResults(playerAIndex) = 1
            payoffResults(playerBIndex) = 1
          else if( (.not.playerCoops(playerAIndex)) .and. &
          &        playerCoops(playerBIndex)) then
            payoffResults(playerAIndex) = 3
            payoffResults(playerBIndex) = 0
          else if( playerCoops(playerAIndex) .and. &
          &        (.not.playerCoops(playerBIndex)) ) then
            payoffResults(playerAIndex) = 0
            payoffResults(playerBIndex) = 3
          end if
        end subroutine

	logical function playerDec(oppCoopLastMove,probCoopIfOppCoop, &
        &                          probCoopIfOppComp, randomPercentages) 
          real          :: probCoopIfOppCoop
          real          :: probCoopIfOppComp
          logical       :: oppCoopLastMove
          integer       :: randomNum 
          real          :: percentProbCoop
          integer       :: randomPercentages

          randomNum = randomPercentages

          if( oppCoopLastMove ) then
            percentProbCoop = ( 100 * (probCoopIfOppCoop) )
          else 
            percentProbCoop = ( 100 * (probCoopIfOppComp) )
          end if

          if( percentProbCoop  .ge. randomNum ) then
            playerDec = .TRUE.
          else 
            playerDec = .FALSE.
          end if

        end function


